python/tank/platform/software_launcher.py

```diff
@@ -0,0 +1,406 @@
+    :returns: :class:`SoftwareLauncher` subclass instance or None.
+    """
+    # Get the engine environment and descriptor using engine.py code
+    (env, engine_descriptor) = get_env_and_descriptor_for_engine(
+        engine_name, tk, context
+    )
+
+    # Make sure it exists locally
+    if not engine_descriptor.exists_local():
+        raise TankError(
+            "Cannot create %s software launcher! %s does not exist on disk." %
+            (engine_name, engine_descriptor)
+        )
+
+    # Get path to engine startup code and load it.
+    engine_path = engine_descriptor.get_path()
+    plugin_file = os.path.join(
+        engine_path, constants.ENGINE_SOFTWARE_LAUNCHER_FILE
+    )
+
+    # Since we don't know what version of the engine is currently
+    # installed, the plugin file may not exist.
+    if not os.path.isfile(plugin_file):
+        # Nothing to do.
+        core_logger.debug(
+            "SoftwareLauncher plugin file '%s' does not exist!" % plugin_file
+        )
+        return None
+
+    core_logger.debug("Loading SoftwareLauncher plugin '%s' ..." % plugin_file)
+    class_obj = load_plugin(plugin_file, SoftwareLauncher)
+    launcher = class_obj(tk, context, engine_name, env)
+    core_logger.debug("Created SoftwareLauncher instance: %s" % launcher)
+
+    # Return the SoftwareLauncher instance
+    return launcher
+
+
+class SoftwareLauncher(object):
+    """
+    Functionality related to the discovery and launch of a DCC
+    """
+    def __init__(self, tk, context, engine_name, env):
+        """
+        Constructor.
+
+        :param tk: :class:`~sgtk.Sgtk` Toolkit instance
+        :param context: :class:`~sgtk.Context` A context object to
+                        define the context on disk where the engine
+                        is operating
+        :param engine_name: (str) Name of the Toolkit engine associated
+                            with the DCC(s) to launch.
+        :param env: An Environment object to associate with this launcher.
+        """
+
+        # get the engine settings
+        settings = env.get_engine_settings(engine_name)
+
+        # get the descriptor representing the engine
+        descriptor = env.get_engine_descriptor(engine_name)
+
+        # check that the context contains all the info that the launcher needs
+        validation.validate_context(descriptor, context)
+
+        # make sure the current operating system platform is supported
+        validation.validate_platform(descriptor)
+
+        # Get the settings for the engine and then validate them
+        engine_schema = descriptor.configuration_schema
+        validation.validate_settings(
+            engine_name,
+            tk,
+            context,
+            engine_schema,
+            settings
+        )
+
+        # Once the engine settings and descriptor have been validated,
+        # initialize members of this class
+        self.__tk = tk
+        self.__context = context
```

Comment:
@manneohrstrom , any concern about caching environment-based data like this when we might be subject to a context change without warning? Not sure if that's anything to worry about here, but worth bringing up.